src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -203,25 +211,66 @@ MakeIntermediateResultHTAB()
 
 /*
  * FindAllWorkerNodesUsingSubplan creates a list of worker nodes that
- * may need to access subplan results.
+ * may need to access subplan results, and decides if these results should be
+ * written to a local file
  */
 List *
-FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
+FindAllWorkerNodesUsingSubplan(IntermediateResultsHashEntry *entry,
 							   char *resultId)
 {
 	List *workerNodeList = NIL;
-	IntermediateResultsHashEntry *entry =
-		SearchIntermediateResult(intermediateResultsHash, resultId);
-
 	ListCell *nodeIdCell = NULL;
 	foreach(nodeIdCell, entry->nodeIdList)
 	{
-		WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+		uint32 nodeId = lfirst_int(nodeIdCell);
+
+		/*
+		 * If we have a dummy placement, intermediate plan will be written locally.
+		 * Note that we do not skip the loop when entry->writeLocalFile, because
+		 * even though the entry is local, it might still be required to broadcast to
+		 * remote nodes.
+		 */
+		if (nodeId == DUMMY_NODE_ID)
+		{
+			entry->writeLocalFile = true;
+			continue;
+		}
+
+		WorkerNode *workerNode = LookupNodeByNodeId(nodeId);
+		Assert(workerNode != NULL);
 
 		workerNodeList = lappend(workerNodeList, workerNode);
+	}
+
+	/* don't include the current worker if the result will be written to local file */
+	if (entry->writeLocalFile)
+	{
+		WorkerNode *workerNode = NULL;
+		int32 localGroupId = GetLocalGroupId();
+
+		/* we'll iterate over the list while deleting from it, so copy it */
+		List *copyOfWorkerNodeList = list_copy(workerNodeList);
+		foreach_ptr(workerNode, copyOfWorkerNodeList)
+		{
+			if (workerNode->groupId == localGroupId)
+			{
+				workerNodeList = list_delete_ptr(workerNodeList, workerNode);
+				break;
+			}
+		}
+	}
+
+	/* now, log the summary */
```

Comment:
Lets expand this comment and explain when we log:
```C
/*
 * Log a summary of decisions made for local execution.
 * By default we log at level DEBUG4
 * When the user has set <name for guc that controls LogIntermediateResults> we change the log level to DEBUG1
 */
```

Also, we have this check here but always log at level `DEBUG1`. This cause the log lines to be prepended with `DEBUG1` even if we log them because we always log them on level `DEBUG4`. But now the user sees a `DEBUG1` message when level is set to `DEBUG4` and not when it is set to `DEBUG1`. This would be super confusing for a user that is debugging an issue.

Instead lets rewrite it to decide the log level separate from the check if the level is loggable:

```C
log_level = DEBUG4;
if (LogIntermediateResults) { log_level = DEBUG1; }
if (IsLoggableLevel(log_level)) {
    ....
    elog(log_level, ....);
    ....
}
```

**edit:** I now see this construct is not introduced in this PR, but I stand my ground around the difference in log level we check for and the log level we log on, huge no-no. So lets address it here anyway since we touch on the logging already in this PR.